### PR TITLE
Fix/1661  better env var support

### PR DESCRIPTION
**Pull request by inkhey**
related to #1661 
description here: https://github.com/tracim/tracim/issues/1661#issuecomment-489103638_

### CODE REVIEW

File: backend/tracim_backend/tests/__init__.py [hunk 1]
```diff
@@ -250,8 +259,14 @@ def override_settings(
         settings["sqlalchemy.url"] = "sqlite://"
         return settings
 
-    def init_database(self, settings: typing.Dict[str, typing.Any]) -> None:
-        self.engine = get_engine(settings)
+    def override_app_config(self) -> None:
```
Comment (lebouquetin) [inline backend/tracim_backend/tests/__init__.py]: what is the usage of this method? need more docstring

File: backend/tracim_backend/tests/__init__.py [hunk 2]
```diff
@@ -120,15 +121,16 @@ def setUp(self) -> None:
         # TODO - G.M - 2019-03-19 - Replace this code by something better, see
         # https://github.com/algoo/hapic/issues/144
         hapic._controllers = []
-        self.connect_database(create_tables=True)
         self.app_config = CFG(self.settings)  # type: CFG
+        self.override_app_config()
```
Comment (lebouquetin) [inline backend/tracim_backend/tests/__init__.py]: I think this is not the way to do this properly, no?

File: backend/tracim_backend/models/setup_models.py [hunk 3]
```diff
@@ -1,26 +1,42 @@
...
+if typing.TYPE_CHECKING:
+    from tracim_backend.config import CFG
 # run configure_mappers after defining all of the models to ensure
 # all relationships can be setup
 configure_mappers()
 
 
-def get_engine(settings, prefix="sqlalchemy."):
-    return engine_from_config(settings, prefix)
+def get_engine(app_config: "CFG", prefix="sqlalchemy."):
```
Comment (lebouquetin) [inline backend/tracim_backend/models/setup_models.py]: typing

File: backend/tracim_backend/models/setup_models.py [hunk 4]
```diff
@@ -1,26 +1,42 @@
...
+from tracim_backend.lib.utils.utils import sliced_dict
 from tracim_backend.models.auth import Group  # noqa: F401
 from tracim_backend.models.auth import Permission  # noqa: F401
 from tracim_backend.models.auth import User  # noqa: F401
 from tracim_backend.models.data import Content  # noqa: F401
 from tracim_backend.models.data import ContentRevisionRO  # noqa: F401
 from tracim_backend.models.meta import DeclarativeBase  # noqa: F401
 
+if typing.TYPE_CHECKING:
+    from tracim_backend.config import CFG
```
Comment (lebouquetin) [inline backend/tracim_backend/models/setup_models.py]: why is the import here? if there is a reason, then a comment is required. If no reason, then move it to top of the file

File: backend/tracim_backend/migration/versions/32e629b17e2e_change_calendar_enabled_to_agenda_enabled_colum.py [hunk 5]
```diff
@@ -5,20 +5,21 @@
...
 
-# revision identifiers, used by Alembic.
 revision = "32e629b17e2e"
 down_revision = "f889c2b59759"
 
 
 def upgrade():
     with op.batch_alter_table("workspaces") as bop:
-        bop.alter_column("calendar_enabled", new_column_name="agenda_enabled")
+        bop.alter_column("calendar_enabled", new_column_name="agenda_enabled", type_=Boolean)
```
Comment (lebouquetin) [inline backend/tracim_backend/migration/versions/32e629b17e2e_change_calendar_enabled_to_agenda_enabled_colum.py]: migrations are not supposed to change. What is the reason for this change?

File: backend/tracim_backend/migration/versions/32e629b17e2e_change_calendar_enabled_to_agenda_enabled_colum.py [hunk 6]
```diff
@@ -5,20 +5,21 @@
...
     # ### end Alembic commands ###
 
 
 def downgrade():
     # ### commands auto generated by Alembic - please adjust! ###
     with op.batch_alter_table("workspaces") as bop:
-        bop.alter_column("agenda_enabled", new_column_name="calendar_enabled")
+        bop.alter_column("agenda_enabled", new_column_name="calendar_enabled", type_=Boolean)
```
Comment (lebouquetin) [inline backend/tracim_backend/migration/versions/32e629b17e2e_change_calendar_enabled_to_agenda_enabled_colum.py]: same remark/question

File: backend/tracim_backend/migration/env.py [hunk 7]
```diff
@@ -14,6 +16,11 @@
 # this is the Alembic Config object, which provides
 # access to the values within the .ini file in use.
 config = context.config
+# INFO - G.M - 2019-05-02 - hack to get env url instead of current config file
+# url for migration script
+env_sqlalchemy_url = os.environ.get("TRACIM_SQLALCHEMY__URL")
+if env_sqlalchemy_url:
+    config.set_main_option("sqlalchemy.url", env_sqlalchemy_url)
```
Comment (lebouquetin) [inline backend/tracim_backend/migration/env.py]: Instead of 

```
env_sqlalchemy_url = os.environ.get("TRACIM_SQLALCHEMY__URL")
if env_sqlalchemy_url:
    config.set_main_option("sqlalchemy.url", env_sqlalchemy_url)
```

we should write this something like :

```
config.set_main_option("sqlalchemy.url", env_sqlalchemy_url, config.get_main_option("sqlalchemy.url")
```

no ?

File: backend/tracim_backend/lib/utils/utils.py [hunk 8]
```diff
@@ -299,3 +299,17 @@ def normpath(path):
     elif path == "":
         path = "/"
     return base_normpath(path)
+
+
+def sliced_dict(data: typing.Dict[str, any], beginning_key_string: str):
```
Comment (lebouquetin) [inline backend/tracim_backend/lib/utils/utils.py]: typing

File: backend/tracim_backend/migration/versions/32e629b17e2e_change_calendar_enabled_to_agenda_enabled_colum.py (hunk 5 above)
Comment (inkhey) [reply backend/tracim_backend/migration/versions/32e629b17e2e_change_calendar_enabled_to_agenda_enabled_colum.py]: i agree. This was needed to fix issue during test of migration for mysql. (this migration is the only one to fail). I do not really know why this parameter "type_" is now needed whereas it was not at all needed before. I can inspect deeper if you think that useful. Anyway this fix, solve the issue.

Comment (lebouquetin) [reply backend/tracim_backend/migration/versions/32e629b17e2e_change_calendar_enabled_to_agenda_enabled_colum.py]: what are we supposed to do on our production? migrate down and up again? Manually patch the database?

File: backend/tracim_backend/migration/env.py (hunk 7 above)
Comment (inkhey) [reply backend/tracim_backend/migration/env.py]: this syntax does not exist for config alembic file.

Comment (lebouquetin) [reply backend/tracim_backend/migration/env.py]: ok.

File: backend/tracim_backend/migration/versions/32e629b17e2e_change_calendar_enabled_to_agenda_enabled_colum.py (hunk 5 above)
Comment (lebouquetin) [reply backend/tracim_backend/migration/versions/32e629b17e2e_change_calendar_enabled_to_agenda_enabled_colum.py]: My feeling is downgrade/upgrade is ok, but I'd like to be sure.

File: backend/tracim_backend/tests/__init__.py (hunk 2 above)
Comment (inkhey) [reply backend/tracim_backend/tests/__init__.py]: you're right, i just refactored it.

File: backend/tracim_backend/tests/__init__.py (hunk 1 above)
Comment (inkhey) [reply backend/tracim_backend/tests/__init__.py]: done.

File: backend/tracim_backend/models/setup_models.py (hunk 3 above)
Comment (inkhey) [reply backend/tracim_backend/models/setup_models.py]: done.

File: backend/tracim_backend/migration/versions/32e629b17e2e_change_calendar_enabled_to_agenda_enabled_colum.py (hunk 5 above)
Comment (inkhey) [reply backend/tracim_backend/migration/versions/32e629b17e2e_change_calendar_enabled_to_agenda_enabled_colum.py]: > what are we supposed to do on our production? migrate down and up again? Manually patch the database?
Nothing. old behaviour does exactly the same. It's just that for unknown reason it's now necessary to specify type_ for mysql. If you want to be sure, i can verify column types from old and new code to exactly match, but i do no see any reason that this method convert column type to another one when we don't specify it (+ if it does something like that, it should probably make many other test failed)

Comment (lebouquetin) [reply backend/tracim_backend/migration/versions/32e629b17e2e_change_calendar_enabled_to_agenda_enabled_colum.py]: ok, thanks for the explanation

File: backend/tracim_backend/models/setup_models.py (hunk 4 above)
Comment (inkhey) [reply backend/tracim_backend/models/setup_models.py]: explained: 'avoid circular import for type-checking usage of CFG' , done.

File: backend/tracim_backend/lib/utils/utils.py (hunk 8 above)
Comment (inkhey) [reply backend/tracim_backend/lib/utils/utils.py]: done

File: backend/tracim_backend/models/data.py [hunk 9]
```diff
@@ -1163,6 +1163,7 @@ def children(self) -> ["Content"]:
             .query(func.max(ContentRevisionRO.revision_id))
             .filter(ContentRevisionRO.content_id.in_(children_content_ids))
             .group_by(ContentRevisionRO.content_id)
+            .order_by(ContentRevisionRO.content_id)
```
Comment (lebouquetin) [inline backend/tracim_backend/models/data.py]: it should be content_revision_id, doesn't it ?

File: backend/tracim_backend/models/data.py [hunk 10]
```diff
@@ -1183,6 +1184,7 @@ def get_children(self, recursively: bool = False) -> ["Content"]:
             children.append(child)
             if recursively:
                 children.extend(child.get_children(recursively=recursively))
+        children = sorted(children, key=lambda child: child.content_id)
```
Comment (lebouquetin) [inline backend/tracim_backend/models/data.py]: same thing here: we should sort by content_id, and by content_revision_id if content_id is the same

File: backend/tracim_backend/models/data.py (hunk 9 above)
Comment (inkhey) [reply backend/tracim_backend/models/data.py]: not here. we sort content not revision here.

File: backend/tracim_backend/models/data.py (hunk 10 above)
Comment (inkhey) [reply backend/tracim_backend/models/data.py]: same answers : we sort content.

**Comment by lebouquetin** [comment]
@inkhey this MR is still WIP? If not, please change the name so that we merge it.

**Comment by inkhey** [comment]
@PhilippeAccorsi , @lebouquetin we merge the associated branch but we do not set the ticket in done, i move in to 2.3, "need test".
